### PR TITLE
[Merge For Next Beta][Analytics] New Schema + messaging / inbox events

### PR DESCRIPTION
**Pull request by mennenia**
YES 🎉 

Long time coming, so much to chat about! Grab a cup of tea 🍵

Tied to [this issue](https://github.com/artsy/collector-experience/issues/446)

This PR is about:

## 1) New typed analytics schema for mobile

Have a look at [track.ts](https://github.com/artsy/emission/blob/mc-msg-analytics/src/lib/utils/track.ts), and you'll see that there's some new content. A few interfaces have been defined for a `Global` track/event schema, as well as a `PageView`. 

### 1.1) PageView information, which is inherited by child components

This is an event that fires on every `componentDidMount`.

The `PageView` interface (defined in `track.ts`) is used in the `AppRegistry` to define a screen view at a high level, and what's nice is that all this content is passed through any event that happens on that screen. 

This is what a pageview looks like (in the console), it fires when a component (the Artist "VC"/root component) mounts:
screen shot 2017-10-30 at 13 09 04

#### What does this mean?

When you follow an artist, you _also_ get all the data about which Artist view you're on. This is the data in the event when you follow an artist (= click the follow button on an artist VC):

screen shot 2017-10-30 at 13 09 14

You can see that all the `context_screen` information carries through. Win!

### 1.2) Schema structure changes

We now will have 3 top level events: **Tap**, **Fail**, and **Success**. 

Imagine sending an inquiry. We capture the fact that the send button has been tapped (Tap event), and then if we get a successful call back from the network, we also fire a Success event under _the same action name_. If something goes wrong, naturally we fire a Fail event instead. 

Sending an inquiry now results in _two_ events.
Output:

screen shot 2017-10-30 at 17 10 23
 
Note that the **`action_name`** property stays the same, but the **`action_type`** is what's different. The idea is that this will be the main event name for the Track events we send through Segment, so that the Analytics team will eventually only have 3 high level tables: Taps, Successes, and Fails. Under those top level categories, you can find the individual events (e.g. `artistFollow`, or `inquirySend`). 

This work (to map the `action_type` to a segement track event, rather than the `action_name` is upcoming and requires Eigen work also). 

### 1.3) Typings give you auto complete for events 🎉 

ezgif-3-772f90d18a

## 2) First pass of Messaging events

I've then used this schema for adding our messaging events (as well as updated the artist follow event. No events yet for Home or Gene — FYI)

### 2.1) Events implemented:

-  conversationSelected: "conversationSelected", (from Inbox list)
-  conversationSendReply: "conversationSendReply", (Tap, Success, and Fail)
-  conversationAttachmentShow: "conversationAttachmentShow", (Tap)
-  conversationAttachmentArtwork: "conversationAttachmentArtwork", (Tap)
-  conversationAttachmentInvoice: "conversationAttachmentInvoice", (Tap)
-  conversationLink: "conversationLinkUsed", (Tap)
-  inquiryCancel: "inquiryCancel", (Tap)
-  inquirySend: "inquirySend", (Tap, Success, and Fail)

### 2.2) Events missing:
- Invoice successfully paid (do we even get a callback? Seeing as we hand off to the web for that cc @mzikherman)
- Zero state inbox of a page view (coming)

--------

### The part where I @ a bunch of people:

Emission crew: @l2succes @sarahscott @mzikherman @alloy 
Just so you have an example of adding new analytics events in the project. Once it's properly finalised and shipped, I can update our docs as well :).

Reaction Proj for GMV: @xtina-starr 
As an example for adding events to the new Onboarding?

Publishing Team: @kanaabe @eessex @craigspaeth 
Adding you as an FYI, I know you've moved to use `react-tracking` in Reaction, and you might be interested in the `track.ts` changes from this PR. If you'd like to add further structure in the one in Reaction, I can definitely take some of that on, if you'd like :)

Analytics: @anipetrov 
So you have a quick glance of the events that I've implemented, and the schema (though we discussed quite a bit over the week). I'll try and get this on device for you ASAP. 

Finally @orta, I think you have the widest overview seeing as @alloy is 🌴, could you review the PR? 🙇‍♀️ 

----

### Coming in another PR:

- Integration with Eigen where the right content gets piped through to the right type of Segment events **DONE, [it's here](https://github.com/artsy/eigen/pull/2425)**

----

Thank you for reading to the end! 💃

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/Components/Inbox/Conversations/__tests__/index-tests.tsx`
- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/lib/utils/track.ts [hunk 1]
```diff
@@ -9,25 +9,118 @@ export namespace Schema {
...
+     *
      */
-    action: string
+    action_type: string
 
     /**
-     * The root container component should specify this as the screen context.
+     * The discription of an event
+     *
+     * E.g. Conversation artwork attatchment tapped
```
Comment (orta) [inline src/lib/utils/track.ts]: tupo

File: src/lib/utils/track.ts [hunk 2]
```diff
@@ -9,25 +9,118 @@ export namespace Schema {
...
+     *
+     * Options are: Tap, Fail, Success
+     *
+     * This is unique to a "Track" event, meaning a "screen view" in Segment does not have this
+     * This is how we distinguish the two type of events in Eigen
+     * Track data inherits the screen view (called "context_screen") properties
+     *
      */
-    action: string
+    action_type: string
```
Comment (orta) [inline src/lib/utils/track.ts]: `action_type: "One" | "Two" | "Three"` describes this better

File: src/lib/utils/track.ts [hunk 3]
```diff
@@ -9,25 +9,118 @@ export namespace Schema {
...
+     */
+    context_screen_owner_id?: string
+
+    /**
+     * The type of entity (owner), E.g. Artist, Artwork, etc.
+     */
+    context_screen_owner_type: string
+  }
+
+  export const PageNames = {
```
Comment (orta) [inline src/lib/utils/track.ts]: enumify all this

Comment (mennenia) [reply src/lib/utils/track.ts]: 🙌 for enums! 

Thanks!

**Comment by mennenia** [comment]
@tayapage6 I know you were curious to have a look! :) I'll link it back in the right cards also

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/Components/Inbox/Conversations/__tests__/index-tests.tsx`
- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/Components/Inbox/Conversations/__tests__/index-tests.tsx`
- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/lib/utils/track.ts (hunk 1 above)
Comment (alloy) [reply src/lib/utils/track.ts]: Best comment ✨

File: src/lib/Components/Inbox/Conversations/Composer.tsx [hunk 4]
```diff
@@ -42,6 +44,9 @@ interface State {
   text: string
 }
 
+const track: Track<Props, State, Schema.Entity> = _track
```
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/Composer.tsx]: Maybe we should make `Schema.Entity` the default schema, seeing as you seem to be defining that in many places?

File: src/lib/utils/track.ts [hunk 5]
```diff
@@ -9,25 +9,118 @@ export namespace Schema {
...
+  export enum OwnerEntityTypes {
+    Artist = "Artist",
+    Artwork = "Artwork",
+    Conversation = "Conversation",
+    Gene = "Gene",
+    Show = "Show",
+    Invoice = "Invoice",
+  }
+
+  export enum ActionEventTypes {
```
Comment (alloy) [inline src/lib/utils/track.ts]: Can this be just `ActionTypes`?

File: src/lib/utils/track.ts [hunk 6]
```diff
@@ -9,25 +9,118 @@ export namespace Schema {
...
+     * Events / results
+     */
+    Fail = "fail",
+    Success = "success",
+  }
+
+  /**
+   * Action event discriptors / names
+   */
+  export enum ActionEventNames {
```
Comment (alloy) [inline src/lib/utils/track.ts]: …and this `ActionEvents`?

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/Components/Inbox/Conversations/__tests__/index-tests.tsx`
- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by mennenia** [comment]
I'll take a look at tests for `track.ts` and also outstanding failures tomorrow.

**Comment by ashfurrow** [comment]
What a nice abstraction!

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by orta** [comment]
Let me know when this is good for a proper review again 👍

**Comment by mennenia** [comment]
@orta good to go! I wasn't sure what other tests you might have had in mind? A lot of what is in in the new track file will be caught by the compiler. 

@alloy added the lowercase test :)

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Missing Test Files:

- `src/lib/utils/__tests__/track-tests.ts`

If these files are supposed to not exist, please update your PR body to include "Skip New Tests".
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by mennenia** [comment]
Can we hold of on merging until the next Eigen beta is cut? This work goes hand in hand with [this Eigen PR](https://github.com/artsy/eigen/pull/2425)

cc @orta @alloy

File: CHANGELOG.md [hunk 7]
```diff
@@ -91,6 +93,7 @@
 -   Added inline rendering for invoices - matt
 -   Removed hard-coded sanitization of message body - sarah
 -   Small composer UI tweaks - maxim
+-   Add analytics events for messaging - maxim
```
Comment (orta) [inline CHANGELOG.md]: these will need to move into the master section now

File: src/lib/Components/Inbox/Conversations/Composer.tsx (hunk 4 above)
Comment (orta) [reply src/lib/Components/Inbox/Conversations/Composer.tsx]: This still needs to be done?

File: src/lib/Containers/Inquiry.tsx [hunk 8]
```diff
@@ -82,6 +83,17 @@ const ResponseRateLine = styled.View`
   margin-top: 5;
 `
 
+// tslint:disable-next-line:no-empty-interface
+interface Props extends RelayProps {}
```
Comment (orta) [inline src/lib/Containers/Inquiry.tsx]: I assume this is a conscious choice to not just throw `RelayProps` below?

Comment (mennenia) [reply src/lib/Containers/Inquiry.tsx]: I think it's mainly consistency, but then an empty interface is also odd. I can remove it.

File: src/lib/Components/Inbox/Conversations/Composer.tsx (hunk 4 above)
Comment (mennenia) [reply src/lib/Components/Inbox/Conversations/Composer.tsx]: So, I don't quite follow... All the interfaces are non default. Do we merge `Entity` with `Global`? Or export the `Entity` as `default`? And then on import, we just use `Schema`? Or can we just import the `Schema` as a whole?

Comment (alloy) [reply src/lib/Components/Inbox/Conversations/Composer.tsx]: The idea is to [here](https://github.com/artsy/emission/blob/08a9039abb7510e1209f996b458f935ae1330131/src/lib/utils/track.ts#L174) change to:

```ts
T extends Schema.Global = Schema.Entity
```

This way the above `_track` cast can be liimted to:

```ts
const track: Track<Props, State> = _track
```

File: src/lib/Components/Inbox/Conversations/Message.tsx [hunk 9]
```diff
@@ -80,7 +80,7 @@ interface Props extends RelayProps {
   conversationId: string
 }
 
-const track: Track<Props, null, Schema.Entity> = _track
+const track: Track<Props, null> = _track
```
Comment (alloy) [inline src/lib/Components/Inbox/Conversations/Message.tsx]: Now that you’ve made `null` the default for `S`, you can remove these as well.

Comment (mennenia) [reply src/lib/Components/Inbox/Conversations/Message.tsx]: 👌

**Comment by mennenia** [comment]
Final final push hopefully :D

**Pull request merged**